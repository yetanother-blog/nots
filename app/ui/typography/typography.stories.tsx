import type { Meta, StoryObj } from '@storybook/react';
import { Typography } from './typography';

/**
 * The Typography component wraps everything you need to render text at Wakesys.
 * It supports headlines, paragraphs, and span elements and some additional
 * options to transform the text. The component comes with some defaults based
 * on the variant. Let's take a look at the variants we have in place:
 */
const meta: Meta<typeof Typography> = {
  title: 'Data Display/Typography',
  component: Typography,
};

export default meta;
type Story = StoryObj<typeof Typography>;

/**
 * ### Default Mapping
 *
 * The goal is to have a smart component in place with some defaults to make it
 * as easy as possible to implement. Therefore, we have implemented a default
 * mapping for the HTML element. It is also possible to override the default
 * mapping by passing the `as` prop (see [polymorphic
 * components](#polymorphic-component-as-prop) section).
 *
 * | **Variant** | **Default HTML Element (Level)** |
 * | ----------- | -------------------------------- |
 * | title       | h1                               |
 * | heading1    | h1                               |
 * | heading2    | h2                               |
 * | heading3    | h3                               |
 * | body        | p                                |
 * | bodySmall   | p                                |
 * | bodyTiny    | p                                |
 */

export const Variants: Story = {
  render: () => (
    <>
      <Typography variant="title">Title</Typography>
      <Typography variant="heading1">Heading 1</Typography>
      <Typography variant="heading2">Heading 2</Typography>
      <Typography variant="heading3">Heading 3</Typography>
      <Typography variant="body">Body</Typography>
      <Typography variant="bodySmall">Body Small</Typography>
    </>
  ),
};

export const Colors: Story = {
  render: () => (
    <>
      <Typography textColor="primary-500" variant="title">
        Title
      </Typography>
      <Typography textColor="primary-400" variant="heading1">
        Heading 1
      </Typography>
      <Typography textColor="primary-300" variant="heading2">
        Heading 2
      </Typography>
      <Typography textColor="grey-800" variant="heading3">
        Heading 3
      </Typography>
      <Typography textColor="grey-700" variant="body">
        Body
      </Typography>
      <Typography textColor="grey-400" variant="bodySmall">
        Body Small
      </Typography>
      <Typography textColor="grey-300" variant="bodyTiny">
        Body Tiny
      </Typography>
    </>
  ),
};

/**
 * We do not encourage this use case, but if you need to change the
 * `font-weight` of the text you can pass the `fontWeight` prop to the
 * component.
 */

export const FontWeight: Story = {
  render: () => (
    <>
      <Typography variant="title" fontWeight="regular" className="mb-4">
        Title with regular font weight
      </Typography>
    </>
  ),
};

export const TextTransform: Story = {
  render: () => (
    <>
      <Typography textTransform="lowercase" variant="heading3">
        Almost before we knew it, we had left the ground.
      </Typography>
      <Typography textTransform="capitalize" variant="heading3">
        Almost before we knew it, we had left the ground.
      </Typography>
      <Typography textTransform="uppercase" variant="heading3">
        Almost before we knew it, we had left the ground.
      </Typography>
    </>
  ),
};

export const TextAlign: Story = {
  render: () => (
    <>
      <Typography textAlign="center" variant="heading3">
        Almost before we knew it, we had left the ground.
      </Typography>
      <Typography textAlign="right" variant="heading3">
        Almost before we knew it, we had left the ground.
      </Typography>
      <Typography textAlign="left" variant="heading3">
        Almost before we knew it, we had left the ground.
      </Typography>
    </>
  ),
};

/**
 * The Typography component supports custom styles by passing the `className`
 * prop to it. Since we use TailwindCSS you can add any utility class to the
 * component you need. In this case we just add some `margin-bottom` to the
 * heading.
 */

export const CustomStyles: Story = {
  parameters: {
    docs: {
      canvas: {
        sourceState: 'shown',
      },
    },
  },
  render: () => (
    <>
      <Typography variant="heading1" className="mb-4">
        Heading 1
      </Typography>
      <Typography>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
        veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea
        commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
        velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint
        occaecat cupidatat non proident, sunt in culpa qui officia deserunt
        mollit anim id est laborum.
      </Typography>
    </>
  ),
};

/**
 * The Typography component is a polymorphic component. This means you can pass
 * any html element to the component and it will render the element you passed.
 * In this case we pass a `label` element to the component.
 */
export const PolymorphicComponentAsProp: Story = {
  parameters: {
    docs: {
      canvas: {
        sourceState: 'shown',
      },
    },
  },
  render: () => (
    <>
      <Typography as="h3" variant="title" className="mb-7">
        Title as H3
      </Typography>
      <Typography
        as="label"
        variant="body"
        textColor="grey-600"
        fontWeight="semibold"
        htmlFor="name"
      >
        Label
      </Typography>
      <input
        className="focus:shadow-outline mb-3 w-full appearance-none rounded border border-ws-danger-500 px-3 py-2 leading-tight text-ws-grey-700 shadow focus:outline-none"
        type="text"
        placeholder="Nothing special..."
        name="name"
        id="name"
      />
    </>
  ),
};
