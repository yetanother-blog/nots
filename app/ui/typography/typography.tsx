import { clsx } from 'clsx';
import type { PolymorphicComponentProp } from '../utils/polymorphic-types';

type Color =
  | 'primary-300'
  | 'primary-400'
  | 'primary-500'
  | 'grey-100'
  | 'grey-200'
  | 'grey-300'
  | 'grey-400'
  | 'grey-500'
  | 'grey-600'
  | 'grey-700'
  | 'grey-800'
  | 'white'
  | 'black';

type Variant =
  | 'title'
  | 'heading1'
  | 'heading2'
  | 'heading3'
  | 'body'
  | 'bodySmall'
  | 'bodyTiny';

type Level = 'h1' | 'h2' | 'h3' | 'p' | 'span';

export type TypographyProps = {
  variant?: Variant;
  textColor?: Color;
  textAlign?: 'left' | 'center' | 'right';
  fontWeight?: 'regular' | 'semibold' | 'bold';
  textTransform?: 'uppercase' | 'capitalize' | 'lowercase';
  className?: string;
};

export const Typography = <C extends React.ElementType = 'p'>({
  children,
  as,
  variant = 'body' as Variant,
  textAlign,
  textColor,
  fontWeight,
  textTransform,
  className,
  ...props
}: PolymorphicComponentProp<C, TypographyProps>) => {
  const titleClasses = clsx(['text-5xl', !fontWeight && 'font-bold']);
  const heading1Classes = clsx([
    'text-3xl md:text-4xl',
    !fontWeight && 'font-bold',
  ]);
  const heading2Classes = clsx([
    'text-xl md:text-3xl',
    !fontWeight && 'font-bold',
  ]);
  const heading3Classes = clsx([
    'text-base md:text-lg',
    !fontWeight && 'font-semibold',
  ]);
  const bodyClasses = clsx(['text-base', !fontWeight && 'font-normal']);
  const bodySmallClasses = clsx(['text-sm', !fontWeight && 'font-normal']);
  const bodyTinyClasses = clsx(['text-xs', !fontWeight && 'font-normal']);

  const textColorClasses = clsx([
    textColor === 'primary-300' && `text-nots-primary-300`,
    textColor === 'primary-400' && `text-nots-primary-400`,
    textColor === 'primary-500' && `text-nots-primary-500`,
    textColor === 'grey-100' && `text-nots-grey-100`,
    textColor === 'grey-200' && `text-nots-grey-200`,
    textColor === 'grey-300' && `text-nots-grey-300`,
    textColor === 'grey-400' && `text-nots-grey-400`,
    textColor === 'grey-500' && `text-nots-grey-500`,
    textColor === 'grey-600' && `text-nots-grey-600`,
    textColor === 'grey-700' && `text-nots-grey-700 dark:text-white`,
    textColor === 'grey-800' && `text-nots-grey-800 dark:text-white`,
    textColor === 'white' && 'text-white',
    textColor === 'black' && 'text-black',
  ]);

  const alignmentClasses = clsx([
    textAlign === 'left' && 'text-left',
    textAlign === 'center' && 'text-center',
    textAlign === 'right' && 'text-right',
  ]);

  const fontWeightClasses = clsx([
    fontWeight === 'regular' && 'font-normal',
    fontWeight === 'semibold' && 'font-semibold',
    fontWeight === 'bold' && 'font-bold',
  ]);

  const textTransformClasses = clsx([
    textTransform === 'uppercase' && 'uppercase',
    textTransform === 'capitalize' && 'capitalize',
    textTransform === 'lowercase' && 'lowercase',
  ]);

  const classes = clsx([
    'antialiased font-sans',
    variant === 'title' && titleClasses,
    variant === 'heading1' && heading1Classes,
    variant === 'heading2' && heading2Classes,
    variant === 'heading3' && heading3Classes,
    variant === 'body' && bodyClasses,
    variant === 'bodySmall' && bodySmallClasses,
    variant === 'bodyTiny' && bodyTinyClasses,
    textColorClasses,
    alignmentClasses,
    fontWeightClasses,
    textTransformClasses,
    className,
  ]);

  const variantsMapping: { [index in Variant]: Level } = {
    title: 'h1',
    heading1: 'h1',
    heading2: 'h2',
    heading3: 'h3',
    body: 'p',
    bodySmall: 'p',
    bodyTiny: 'p',
  };

  const Component = as ?? variantsMapping[variant];
  return (
    <Component className={classes} {...props}>
      {children}
    </Component>
  );
};
