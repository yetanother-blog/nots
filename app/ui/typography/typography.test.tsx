import { render } from '@testing-library/react';
import { Typography } from './typography';

describe('Typography Component', () => {
  describe('Variants with default mapping`', () => {
    it('should render with variant title as h1', () => {
      const { container } = render(
        <Typography variant="title">Test</Typography>
      );
      expect(container).toBeInTheDocument();

      const element = container.querySelector('h1');
      expect(element).toBeInTheDocument();
    });

    it('should render with variant heading1 as h1', () => {
      const { container } = render(
        <Typography variant="heading1">Test</Typography>
      );
      expect(container).toBeInTheDocument();

      const element = container.querySelector('h1');
      expect(element).toBeInTheDocument();
    });

    it('should render with variant heading2 as h2', () => {
      const { container } = render(
        <Typography variant="heading2">Test</Typography>
      );
      expect(container).toBeInTheDocument();

      const element = container.querySelector('h2');
      expect(element).toBeInTheDocument();
    });

    it('should render with variant heading3 as h3', () => {
      const { container } = render(
        <Typography variant="heading3">Test</Typography>
      );
      expect(container).toBeInTheDocument();

      const element = container.querySelector('h3');
      expect(element).toBeInTheDocument();
    });

    it('should render with variant body as p', () => {
      const { container } = render(
        <Typography variant="body">Test</Typography>
      );

      expect(container).toBeInTheDocument();

      const element = container.querySelector('p');
      expect(element).toBeInTheDocument();
    });

    it('should render with variant bodySmall as p', () => {
      const { container } = render(
        <Typography variant="bodySmall">Test</Typography>
      );
      expect(container).toBeInTheDocument();

      const element = container.querySelector('p');
      expect(element).toBeInTheDocument();
    });

    it('should render with variant bodyTiny as p', () => {
      const { container } = render(
        <Typography variant="bodyTiny">Test</Typography>
      );
      expect(container).toBeInTheDocument();

      const element = container.querySelector('p');
      expect(element).toBeInTheDocument();
    });
  });
  describe('as prop`', () => {
    it('should accept label props', () => {
      const { container } = render(
        <Typography variant="title" as="label" htmlFor="name">
          Test
        </Typography>
      );
      expect(container).toBeInTheDocument();

      const element = container.querySelector('label');
      const htmlFor = element?.getAttribute('for');

      expect(htmlFor).toBe('name');
      expect(element).toBeInTheDocument();
    });
  });

  describe('font props', () => {
    it('should render with text transform uppercase', () => {
      const { container } = render(
        <Typography variant="title" textTransform="uppercase">
          Test
        </Typography>
      );
      expect(container).toMatchSnapshot();
    });
    it('should render with text transform capitalize', () => {
      const { container } = render(
        <Typography variant="title" textTransform="capitalize">
          Test
        </Typography>
      );
      expect(container).toMatchSnapshot();
    });
    it('should render with text transform lowercase', () => {
      const { container } = render(
        <Typography variant="title" textTransform="lowercase">
          Test
        </Typography>
      );
      expect(container).toMatchSnapshot();
    });
    it('should render with text align center', () => {
      const { container } = render(
        <Typography variant="title" textAlign="center">
          Test
        </Typography>
      );
      expect(container).toMatchSnapshot();
    });
    it('should render with text align right', () => {
      const { container } = render(
        <Typography variant="title" textAlign="right">
          Test
        </Typography>
      );
      expect(container).toMatchSnapshot();
    });
    it('should render with text align left', () => {
      const { container } = render(
        <Typography variant="title" textAlign="left">
          Test
        </Typography>
      );
      expect(container).toMatchSnapshot();
    });
  });
});
