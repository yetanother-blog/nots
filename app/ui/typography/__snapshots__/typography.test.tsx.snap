// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Typography Component > font props > should render with text align center 1`] = `
<div>
  <h1
    class="antialiased font-sans text-5xl font-bold text-center"
  >
    Test
  </h1>
</div>
`;

exports[`Typography Component > font props > should render with text align left 1`] = `
<div>
  <h1
    class="antialiased font-sans text-5xl font-bold text-left"
  >
    Test
  </h1>
</div>
`;

exports[`Typography Component > font props > should render with text align right 1`] = `
<div>
  <h1
    class="antialiased font-sans text-5xl font-bold text-right"
  >
    Test
  </h1>
</div>
`;

exports[`Typography Component > font props > should render with text transform capitalize 1`] = `
<div>
  <h1
    class="antialiased font-sans text-5xl font-bold capitalize"
  >
    Test
  </h1>
</div>
`;

exports[`Typography Component > font props > should render with text transform lowercase 1`] = `
<div>
  <h1
    class="antialiased font-sans text-5xl font-bold lowercase"
  >
    Test
  </h1>
</div>
`;

exports[`Typography Component > font props > should render with text transform uppercase 1`] = `
<div>
  <h1
    class="antialiased font-sans text-5xl font-bold uppercase"
  >
    Test
  </h1>
</div>
`;
